cachi2/core/package_managers/gomod.py

```diff
@@ -88,29 +88,42 @@ def fetch_gomod_source(request: Request) -> RequestOutput:
 
     packages = []
 
-    for i, subpath in enumerate(subpaths):
-        log.info("Fetching the gomod dependencies at subpath %s", subpath)
+    with GoCacheTemporaryDirectory(prefix="cachito-") as tmp_dir:
+        for i, subpath in enumerate(subpaths):
+            log.info("Fetching the gomod dependencies at subpath %s", subpath)
 
-        log.info(f'Fetching the gomod dependencies at the "{subpath}" directory')
+            log.info(f'Fetching the gomod dependencies at the "{subpath}" directory')
 
-        gomod_source_path = request.source_dir / subpath
-        try:
-            gomod = _resolve_gomod(gomod_source_path, request)
-        except GoModError:
-            log.error("Failed to fetch gomod dependencies")
-            raise
+            gomod_source_path = request.source_dir / subpath
+            try:
+                gomod = _resolve_gomod(gomod_source_path, request, tmp_dir)
+            except GoModError:
+                log.error("Failed to fetch gomod dependencies")
+                raise
 
-        module_info = gomod["module"]
+            module_info = gomod["module"]
 
-        packages.append({**module_info, "path": subpath, "dependencies": gomod["module_deps"]})
+            packages.append({**module_info, "path": subpath, "dependencies": gomod["module_deps"]})
 
-        # add package deps
-        for package in gomod["packages"]:
-            pkg_info = package["pkg"]
-            package_subpath = _package_subpath(module_info["name"], pkg_info["name"], subpath)
-            packages.append(
-                {**pkg_info, "path": package_subpath, "dependencies": package.get("pkg_deps", [])}
-            )
+            # add package deps
+            for package in gomod["packages"]:
+                pkg_info = package["pkg"]
+                package_subpath = _package_subpath(module_info["name"], pkg_info["name"], subpath)
+                packages.append(
+                    {
+                        **pkg_info,
+                        "path": package_subpath,
+                        "dependencies": package.get("pkg_deps", []),
+                    }
+                )
+
+        tmp_download_cache_dir = os.path.join(tmp_dir, request.go_mod_cache_download_part)
+        log.debug(
+            "Adding dependencies from %s to %s",
+            tmp_download_cache_dir,
+            request.gomod_download_dir,
+        )
+        shutil.copytree(tmp_download_cache_dir, str(request.gomod_download_dir), dirs_exist_ok=True)
```

Comment:
We should copy the cache only if we're not vendoring (that's what's causing most of the integration test failures)